components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/DynamicX509TrustManager.java

```diff
@@ -0,0 +1,141 @@
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+package org.wso2.carbon.identity.core.util;
+
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.base.ServerConfiguration;
+import org.wso2.carbon.base.api.ServerConfigurationService;
+
+
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509TrustManager;
+import java.io.*;
+import java.security.KeyStoreException;
+import java.security.KeyStore;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+
+
+/**
+ * Gives a singleton javax.net.ssl.X509TrustManager implementation that uses the default carbon trust store.
+ * This will load any changes (addition/removal of certificates) done to the default trust store on the fly.
+ */
+public class DynamicX509TrustManager implements X509TrustManager {
+    private static Log log = LogFactory.getLog(DynamicX509TrustManager.class);
+    private X509TrustManager trustManager;
+    private static DynamicX509TrustManager instance;
+    //System Properties
+    public static final String PROP_TRUST_STORE_UPDATE_REQUIRED = "org.wso2.carbon.identity.core.util.DynamicX509TrustManager.TRUST_STORE_UPDATE_REQUIRED";
+
+    private DynamicX509TrustManager() throws Exception{
+        setupTrustManager();
+    }
+
+    public static DynamicX509TrustManager getInstance() throws Exception {
+        if(instance == null)
+            instance = new DynamicX509TrustManager();
+        return instance;
+    }
+    @Override
+    public void checkClientTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {
+        trustManager.checkClientTrusted(x509Certificates, s);
+    }
+
+    /**
+     * Checks the validity of passed x509Certificate certificate chain
+     * @param x509Certificates
+     * @param s
+     * @throws CertificateException
+     */
+    @Override
+    public void checkServerTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {
+        try {
+            //if changes were made to the trust store, reload the trust store and initialize the trustManager instance.
+            if("true".equals(System.getProperty(PROP_TRUST_STORE_UPDATE_REQUIRED))) {
+                setupTrustManager();
+            }
+            trustManager.checkServerTrusted(x509Certificates, s);
+        } catch (NoSuchAlgorithmException e) {
+            throw new CertificateException("Certificate validation failed due to " + e.getCause());
+        } catch (KeyStoreException e) {
+            throw new CertificateException("Certificate validation failed due to " + e.getCause());
+        } catch (Exception e) {
+            throw new CertificateException("Certificate validation failed due to " + e.getCause());
+        }
```

Comment:
use
throw new IdentityException ("Meaningful message", e),
 its better to use different messages that explain the reason for each exception at different catch blocks .